runtime/gc_vlhgc/SweepPoolManagerVLHGC.cpp

```diff
@@ -50,289 +50,61 @@ MM_SweepPoolManagerVLHGC::newInstance(MM_EnvironmentBase *env)
-	memoryPool->incrementScannableBytes(chunk->_scannableBytes, chunk->_nonScannableBytes);
-	Assert_MM_true((sweepState->_sweepFreeBytes + memoryPool->getDarkMatterBytes()) <= _extensions->regionSize);
-}
-
-void
-MM_SweepPoolManagerVLHGC::flushFinalChunk(MM_EnvironmentBase *envModron, MM_MemoryPool *memoryPool)
-{
-	MM_SweepPoolState *sweepState = getPoolState(memoryPool);
-
-	/* If the last chunk had trailing free space, try and add it to the free list */
-	if((sweepState->_connectPreviousChunk != NULL) && (sweepState->_connectPreviousChunk->trailingFreeCandidateSize > 0)) {
-		/* Check if the entry is a candidate */
-		if (((MM_MemoryPoolBumpPointer *)memoryPool)->canMemoryBeConnectedToPool(envModron, sweepState->_connectPreviousChunk->trailingFreeCandidate, sweepState->_connectPreviousChunk->trailingFreeCandidateSize)) {
-			sweepState->_connectPreviousFreeEntry = (MM_HeapLinkedFreeHeader *)sweepState->_connectPreviousChunk->trailingFreeCandidate;
-			sweepState->_connectPreviousFreeEntrySize = sweepState->_connectPreviousChunk->trailingFreeCandidateSize;
-			Assert_MM_true(sweepState->_connectPreviousFreeEntry != sweepState->_connectPreviousChunk->leadingFreeCandidate);
-	
-			sweepState->_sweepFreeBytes += sweepState->_connectPreviousChunk->trailingFreeCandidateSize;
-			sweepState->_sweepFreeHoles += 1;
-			sweepState->_largestFreeEntry = OMR_MAX(sweepState->_connectPreviousChunk->trailingFreeCandidateSize, sweepState->_largestFreeEntry);
-		}
+	if (0 != adjustedbytes) {
+		((MM_MemoryPoolAddressOrderedList *)memoryPool)->setAdjustedBytesForCardAlignment(adjustedbytes, needSync);
 	}
 }
 
-void
-MM_SweepPoolManagerVLHGC::connectFinalChunk(MM_EnvironmentBase *envModron,MM_MemoryPool *memoryPool)
+MMINLINE bool
+MM_SweepPoolManagerVLHGC::isEligibleForFreeMemory(MM_EnvironmentBase *env, MM_MemoryPoolAddressOrderedListBase *memoryPool, void* address, uintptr_t size)
 {
-	/* Update pool free memory statistics since they are required to identify free regions to recycle */
-	MM_SweepPoolState *sweepState = getPoolState(memoryPool);
-	memoryPool->setFreeMemorySize(sweepState->_sweepFreeBytes);
-	memoryPool->setFreeEntryCount(sweepState->_sweepFreeHoles);
-	memoryPool->setLargestFreeEntry(sweepState->_largestFreeEntry);
-	MM_MemoryPoolBumpPointer *bpPool = (MM_MemoryPoolBumpPointer *)memoryPool;
-	UDATA actualFreeMemory = bpPool->getActualFreeMemorySize();
-	UDATA allocatableBytes = bpPool->getAllocatableBytes();
-	if (0 == actualFreeMemory) {
-		Assert_MM_true(allocatableBytes < bpPool->getMinimumFreeEntrySize());
+	bool ret = false;
+	if (!env->_cycleState->_noCompactionAfterSweep)
+	{
+		ret = memoryPool->canMemoryBeConnectedToPool(env, address, size);
 	} else {
-		Assert_MM_true(allocatableBytes <= actualFreeMemory);
+		uintptr_t minimumSize = OMR_MAX(((MM_GCExtensions*)_extensions)->minimumFreeSizeForSurvivor, memoryPool->getMinimumFreeEntrySize());
+		if (size >= minimumSize) {
+			ret = true;
+		}
```

Comment:
could we do this computation only once at the start of sweep phase, and store minFreeSize in SweepPoolManager, and here just compare against that size?

Reply:
We could probably live without isEligibleForFreeMemory being virtual either. That minFreeSize could be part of base SweepPoolManager, and preprocessing step of sweep phase would be virtual.

On the other side, in general case, we don't know what canMemoryBeConnectedToPool does - it could be doing different things for different positions of free entry (different addresses), but in fact it only looks at the size and has one simple size threshold to compare with. But I'd be ok to rely directly on MP's min free size directly, and get rid of canMemoryBeConnectedToPool  API altogether. 

But, wait for now - I would not delay this PR just for this improvement (obviously it would more OMR changes)....